.travis.yml

```diff
@@ -1,30 +1,111 @@
   - sudo dpkg -i mysql-apt-config_0.8.13-1_all.deb
+  # Percona tools
+  - wget https://repo.percona.com/apt/percona-release_latest.$(lsb_release -sc)_all.deb
   - sudo dpkg -i percona-release_latest.$(lsb_release -sc)_all.deb
   - sudo percona-release enable-only tools
-  - sudo apt-get update -q
-  - sudo apt-get install -q -y --allow-unauthenticated -o Dpkg::Options::=--force-confnew mysql-server
-  - sudo apt-get install percona-xtrabackup-80
+  - sudo apt update
+
+
+inc_install: &inc_install
+  # Build Deps
+  - sudo apt install -y libsnappy-dev
+  - sudo apt-get install -q -y --allow-unauthenticated -o Dpkg::Options::=--force-confnew mysql-server percona-xtrabackup-80
+  # Setup common python dependencies
+  - pip install -r requirements.txt
+  - pip install -r requirements.dev.txt
+  - pip install -e .
+
+inc_unittest: &inc_unittest
+  - make unittest
+
+jobs:
+  allow_failures:
+    - stage: lint
+      python: 3.6
+      env: [MYSQL_VERSION=mysql-5.7]
+      name: fmt
+      before_install: *inc_before_install
+      install: *inc_install
+      script:
+        - make fmt
+        - if [ $(git diff --name-only --diff-filter=ACMR | wc -l ) != 0 ]; then
+          echo "Reformatting failed! Please run make fmt on your commits and resubmit!" 1>&2 ;
+          git diff ;
+          exit 1 ;
+          fi
+
+    - stage: typecheck
+      python: 3.6
+      env: [MYSQL_VERSION=mysql-5.7]
+      name: fmt
+      before_install: *inc_before_install
+      install: *inc_install
+      script:
+        - make typecheck
+  include:
+    # Run linters with our "primary" Python version.
+    - stage: lint
+      python: 3.6
+      env: [MYSQL_VERSION=mysql-5.7]
+      name: lint
+      before_install: *inc_before_install
+      install: *inc_install
+      script:
+        - make lint
+
+    # Matrix build unittests
+    - env: [MYSQL_VERSION=mysql-5.7]
+      python: '3.6'
+      stage: unittest
+      before_install: *inc_before_install
+      install: *inc_install
+      script: *inc_unittest
+
+    - env: [MYSQL_VERSION=mysql-5.7]
+      python: '3.6'
+      stage: unittest
+      before_install: *inc_before_install
+      install: *inc_install
+      script: *inc_unittest
+
+    - env: [MYSQL_VERSION=mysql-5.7]
+      python: '3.7'
+      stage: unittest
+      before_install: *inc_before_install
+      install: *inc_install
+      script: *inc_unittest
+
+    - env: [MYSQL_VERSION=mysql-5.7]
+      python: '3.7'
+      stage: unittest
+      before_install: *inc_before_install
+      install: *inc_install
+      script: *inc_unittest
 
-script:
-  - make test
+    - env: [MYSQL_VERSION=mysql-5.7]
+      python: '3.8'
+      stage: unittest
+      before_install: *inc_before_install
+      install: *inc_install
+      script: *inc_unittest
 
+    - env: [MYSQL_VERSION=mysql-5.7]
+      python: '3.8'
+      stage: unittest
+      before_install: *inc_before_install
+      install: *inc_install
+      script: *inc_unittest
```

Comment:
Haven't been configuring Travis a lot so maybe I'm missing something but does defining exactly the same entry twice for each Python version have some effect or do these have some difference that I'm not seeing?

Reply:
That is a mistake. So originally we had MYSQL 5.6, 5.7 and 8.0. Let me reinstate some of the other versions and we'll see if the full matrix works (and it'll fix up this problem).